pkg/proxy/iptables/proxier.go

```diff
@@ -425,32 +439,92 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
 			info.loadBalancerStatus = *api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
 			info.sessionAffinityType = service.Spec.SessionAffinity
+			if l, ok := service.Annotations[apiservice.AnnotationHandleExternalTraffic]; ok {
+				if l == apiservice.AnnotationValueExternalEndpointsLocal {
+					info.onlyNodeLocalEndpoints = true
+				} else if l == apiservice.AnnotationValueExternalEndpointsGlobal {
+					info.onlyNodeLocalEndpoints = false
+				} else {
+					glog.Errorf("Invalid value for annotation %v", apiservice.AnnotationHandleExternalTraffic)
+					info.onlyNodeLocalEndpoints = false
+				}
+				if l, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; ok {
+					p, err := strconv.Atoi(l)
+					if err != nil {
+						glog.Errorf("Failed to parse annotation %v: %v",
+							apiservice.AnnotationHealthCheckNodePort, err)
+					}
+					info.healthCheckNodePort = p
+					healthcheck.AddServiceListener(serviceName.NamespacedName, info.healthCheckNodePort)
+				} else {
+					glog.Errorf("Service does not contain necessary annotation %v",
+						apiservice.AnnotationHealthCheckNodePort)
+				}
+			}
+
 			proxier.serviceMap[serviceName] = info
 
 			glog.V(4).Infof("added serviceInfo(%s): %s", serviceName, spew.Sdump(info))
 		}
 	}
 
 	staleUDPServices := sets.NewString()
-	// Remove services missing from the update.
-	for name := range proxier.serviceMap {
+	activeServiceListeners := sets.NewString()
+	for name := range activeServices {
+		activeServiceListeners.Insert(name.NamespacedName.String())
+	}
+	// Remove serviceports missing from the update.
+	for name, info := range proxier.serviceMap {
 		if !activeServices[name] {
 			glog.V(1).Infof("Removing service %q", name)
-			if proxier.serviceMap[name].protocol == api.ProtocolUDP {
-				staleUDPServices.Insert(proxier.serviceMap[name].clusterIP.String())
+			if info.protocol == api.ProtocolUDP {
+				staleUDPServices.Insert(info.clusterIP.String())
 			}
 			delete(proxier.serviceMap, name)
 		}
+		if info.onlyNodeLocalEndpoints && info.healthCheckNodePort > 0 {
+			// Remove ServiceListener health check nodePorts from the health checker
+			if !activeServiceListeners.Has(name.NamespacedName.String()) {
+				healthcheck.DeleteServiceListener(name.NamespacedName, info.healthCheckNodePort)
+			}
+		}
 	}
-
 	proxier.syncProxyRules()
 	proxier.deleteServiceConnections(staleUDPServices.List())
 
 }
 
+func flattenEndpointsInfo(endPoints []*endpointsInfo) []string {
+	var endpointIPs []string
+	for _, ep := range endPoints {
+		endpointIPs = append(endpointIPs, ep.ip)
+	}
+	return endpointIPs
+}
+
+func (proxier *Proxier) filterEndpoints(endPoints []hostPortInfo, endpointIPs []string) []*endpointsInfo {
+
+	lookupMap := make(map[string]bool)
+	for _, ip := range endpointIPs {
+		lookupMap[ip] = true
+	}
+
+	var filteredEndpoints []*endpointsInfo
+	for _, hpp := range endPoints {
+		key := net.JoinHostPort(hpp.host, strconv.Itoa(hpp.port))
```

Comment:
Reading this I am confused.  lookupMap is derived from  from endpointIPs, so I assume it is keyed by IP strings.  But here I see you generating a key that is ip:port.  So SOMETHING is named badly and under-commented.  This code is tricky.  Please err on the side of too much explanation.